client/fingerprint/consul.go

```diff
@@ -173,11 +182,50 @@ func (f *ConsulFingerprint) connect(info agentconsul.Self) (string, bool) {
 	return strconv.FormatBool(c), ok
 }
 
-func (f *ConsulFingerprint) grpc(info agentconsul.Self) (string, bool) {
+func (f *ConsulFingerprint) grpc(scheme string) func(info agentconsul.Self) (string, bool) {
+	return func(info agentconsul.Self) (string, bool) {
+
+		// The version is needed in order to understand which config object to
+		// query. This is because Consul 1.14.0 added a new gRPC port which
+		// broke the previous behaviour.
+		v, ok := info["Config"]["Version"].(string)
+		if !ok {
+			return "", false
+		}
+
+		consulVersion, err := version.NewVersion(v)
+		if err != nil {
+			return "", false
+		}
+
+		// If the Consul agent being fingerprinted is running a version less
+		// than 1.14.0 we use the original single gRPC port.
+		if consulVersion.Core().LessThan(consulGRPCPortChangeVersion.Core()) {
+			return f.grpcPort(info)
+		}
+
+		// Now that we know we are querying a Consul agent running v1.14.0 or
+		// greater, we need to select the correct port parameter from the
+		// config depending on whether we have been asked to speak TLS or not.
+		switch scheme {
+		case "https", "HTTPS":
```

Comment:
Not a blocker, but I think lowering (or raising) is a better approach generally:

```suggestion
		switch strings.ToLower(scheme) {
		case "https":
```

Reply:
Agreed; I was trying to be clever and avoid the func overhead but I can now see what can go wrong with that approach. This has been fixed in commit: https://github.com/hashicorp/nomad/pull/15309/commits/497eb98b161768cb8ccc54e9a3f577d90908470e